src/outputs/parthenon_hdf5.cpp

```diff
@@ -15,17 +15,158 @@
 //========================================================================================
 
+// options for building
+#include "config.hpp"
+
 // Only proceed if HDF5 output enabled
+#ifdef HDF5OUTPUT
 
+#include <algorithm>
 #include <memory>
+#include <set>
+#include <type_traits>
+#include <unordered_map>
 
+#include "interface/meshblock_data_iterator.hpp"
+#include "interface/metadata.hpp"
+#include "mesh/mesh.hpp"
+#include "mesh/meshblock.hpp"
+#include "outputs/outputs.hpp"
 #include "outputs/parthenon_hdf5.hpp"
+#include "utils/string_utils.hpp"
 
-#include "mesh/meshblock.hpp"
+namespace parthenon {
+namespace HDF5 {
+
+// template specializations for std::string
+template <>
+void HDF5WriteAttribute(const std::string &name, const std::vector<std::string> &values,
+                        hid_t location) {
+  std::vector<const char *> char_ptrs(values.size());
+  for (size_t i = 0; i < values.size(); ++i) {
+    char_ptrs[i] = values[i].c_str();
+  }
+  HDF5WriteAttribute(name, char_ptrs, location);
+}
 
-#ifdef HDF5OUTPUT
+template <>
+std::vector<std::string> HDF5ReadAttributeVec(hid_t location, const std::string &name) {
+  // get strings as char pointers, HDF5 will allocate the memory and we need to free it
+  auto char_ptrs = HDF5ReadAttributeVec<char *>(location, name);
 
-namespace parthenon {
+  // make strings out of char pointers, which copies the memory and then free the memeory
+  std::vector<std::string> res(char_ptrs.size());
+  for (size_t i = 0; i < res.size(); ++i) {
+    res[i] = std::string(char_ptrs[i]);
+    free(char_ptrs[i]);
+  }
+
+  return res;
+}
+
+// template specialization for bool
+template <>
+void HDF5WriteAttribute(const std::string &name, const std::vector<bool> &values,
+                        hid_t location) {
+  // can't use std::vector here because std::vector<bool>  doesn't have .data() member
+  std::unique_ptr<hbool_t[]> data(new hbool_t[values.size()]);
+  for (size_t i = 0; i < values.size(); ++i) {
+    data[i] = values[i];
+  }
+  HDF5WriteAttribute(name, values.size(), data.get(), location);
+}
+
+} // namespace HDF5
+
+using namespace HDF5;
+
+// Helper struct containing some information about a variable that we can easily
+// communicate via MPI
+struct VarInfo {
+  // We need to communicate this struct via MPI. To Make our lives a bit easier, we will
+  // combine the vlen integer and the is_sparse and is_vector flags into a single int
+  // (call it info_code) and communicate that.
+  //
+  // The info_code will have the vlen in the lower 16 bits and bits 20 and 21 will encode
+  // the is_sparse and is_vector flags
+  static constexpr int max_vlen = (1 << 16) - 1;
+  static constexpr int sparse_flag = (1 << 20);
+  static constexpr int vector_flag = (1 << 21);
+
+  std::string label;
+  std::vector<std::string> component_labels; // may be empty
+  int vlen;
+  bool is_sparse;
+  bool is_vector;
+
+ private:
+  VarInfo() = default;
```

Comment:
The reason is because `VarInfo` should not default-constructed because then it would not be initialized. But I was too lazy to add a general constructor and need to initialize the fields individually in the `Decode` static constructor. But I should probably add a proper constructor and use that.